src/broadcaster.rs

```diff
@@ -0,0 +1,192 @@
+            shared_state: self.state.shared_state.clone()
+        });
+
+        Self {
+            state: new_state,
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------------------------
+
+// State for a single broadcaster instance.
+//
+// It's split in this way because the BroadcasterSharedState also needs to access the status
+// (e.g. to notify of changes).
+struct BroadcasterState<A: Signal> {
+    status: Arc<Mutex<BroadcasterStatus>>,
+    shared_state: Arc<Mutex<BroadcasterSharedState<A>>>,
+}
+
+// ---------------------------------------------------------------------------------------------
+
+struct BroadcasterStatus {
+    has_changed: bool,
+    waker: Option<Waker>
+}
+
+// ---------------------------------------------------------------------------------------------
+
+// Shared state underpinning a Cloned set of broadcasters
+struct BroadcasterSharedState<A: Signal> {
+    signal: A,
+    value: Option<A::Item>,
+    notifier: Arc<BroadcasterNotifier>
+}
+
+impl<A: Signal> BroadcasterSharedState<A>
+        where <A as Signal>::Item: Clone {
+
+    // Poll the underlying signal for changes, giving it a BroadcasterWaker to wake in
+    // the future if it is in Pending state.
+    //
+    // Takes self by Arc<Mutex<Self>> because this construct is needed to build a
+    // BroadcasterWaker
+    fn poll_change(&mut self, cx: &mut Context) -> Async<Option<A::Item>> {
+
+        // Set the notifier to be waiting on a poll, and if it wasn't previously, actually
+        // execute that poll.
+        //
+        // TODO: Do we actually need AcqRel guarantee?
+        if !self.notifier.is_waiting.swap(true, Ordering::AcqRel) {
+            let waker = Waker::from(self.notifier.clone());
+            let mut sub_cx = cx.with_waker(&waker);
+
+            if let Async::Ready(value) = self.signal.poll_change(&mut sub_cx) {
+                self.value = value.clone();
+                self.notifier.notify();
+                self.notifier.is_waiting.store(false, Ordering::Release);
+                return Async::Ready(value)
+            }
+        }
+
+        Async::Pending
+    }
+}
+
+// ----------------------------------------------------------------------------------------------
+
+/// This is responsible for propagating a "wake" down to any pending tasks attached to
+/// broadcasted children.
+struct BroadcasterNotifier {
+    is_waiting: AtomicBool,
+    targets: Mutex<Vec<Weak<Mutex<BroadcasterStatus>>>>
+}
+
+impl BroadcasterNotifier {
+    fn notify(&self) {
+        // Take this opportunity to GC dead children
+        self.targets.lock().unwrap().retain(|weak_child_state| {
+            if let Some(child_status_mutex) = weak_child_state.upgrade() {
+                let mut child_status = child_status_mutex.lock().unwrap();
+                child_status.has_changed = true;
+
+                if let Some(waker) = child_status.waker.take() {
+                    waker.wake();
```

Comment:
Shouldn't this `drop` the lock before it wakes it? I know Executors are supposed to notify asynchronously, but I still like the idea of dropping the lock ASAP.